components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java

```diff
@@ -0,0 +1,366 @@
+        this.keyStorePath = keyStorePath;
+    }
+
+    public String getKeyStorePassword() {
+        return keyStorePassword;
+    }
+
+    public void setKeyStorePassword(String keyStorePassword) {
+        this.keyStorePassword = keyStorePassword;
+    }
+
+    public String getDialectURI() {
+        return dialectURI;
+    }
+
+    public void setDialectURI(String dialectURI) {
+        this.dialectURI = dialectURI;
+    }
+
+    public String getSignatureAlgorithm() {
+        return signatureAlgorithm;
+    }
+
+    public void setSignatureAlgorithm(String signatureAlgorithm) {
+        this.signatureAlgorithm = signatureAlgorithm;
+    }
+
+    public int getJwtExpiryTime() {
+        return jwtExpiryTime;
+    }
+
+    public void setJwtExpiryTime(int jwtExpiryTime) {
+        this.jwtExpiryTime = jwtExpiryTime;
+    }
+
+    /**
+     * Used to generate the JWT token
+     */
+    public String generateToken(Map<String, Object> jwtInfo) throws Exception {
+        String jwtHeader = buildHeader();
+        String jwtBody = buildBody(jwtInfo);
+        String base64UrlEncodedHeader = "";
+        if (jwtHeader != null) {
+            base64UrlEncodedHeader = encode(jwtHeader.getBytes(Charset.defaultCharset()));
+        }
+        String base64UrlEncodedBody = "";
+        if (jwtBody != null) {
+            base64UrlEncodedBody = encode(jwtBody.getBytes());
+        }
+        if (SHA256_WITH_RSA.equals(signatureAlgorithm)) {
+            String assertion = base64UrlEncodedHeader + '.' + base64UrlEncodedBody;
+            //get the assertion signed
+            byte[] signedAssertion = signJWT(assertion);
+            if (logger.isDebugEnabled()) {
+                logger.debug("signed assertion value : " + new String(signedAssertion, Charset.defaultCharset()));
+            }
+            String base64UrlEncodedAssertion = encode(signedAssertion);
+            return base64UrlEncodedHeader + '.' + base64UrlEncodedBody + '.' + base64UrlEncodedAssertion;
+        } else {
+            return base64UrlEncodedHeader + '.' + base64UrlEncodedBody + '.';
+        }
+    }
+
+    /**
+     * Used to build the JWT header
+     */
+    public String buildHeader() throws Exception {
+        String jwtHeader = null;
+        if (NONE.equals(signatureAlgorithm)) {
+            StringBuilder jwtHeaderBuilder = new StringBuilder();
+            jwtHeaderBuilder.append("{\"typ\":\"JWT\",");
+            jwtHeaderBuilder.append("\"alg\":\"");
+            jwtHeaderBuilder.append("none");
+            jwtHeaderBuilder.append('\"');
+            jwtHeaderBuilder.append('}');
+
+            jwtHeader = jwtHeaderBuilder.toString();
+
+        } else if (SHA256_WITH_RSA.equals(signatureAlgorithm)) {
+            jwtHeader = addCertToHeader();
+        }
+        return jwtHeader;
+    }
+
+    /**
+     * Used to sign the JWT using the keystore
+     */
+    public byte[] signJWT(String assertion) throws Exception {
+        FileInputStream is;
```

Comment:
better to add some debug logs in each step of JWT generation to find out where it fails if we notice an error in generation.